element_array_ephys/export/nwb/nwb.py

```diff
@@ -0,0 +1,600 @@
+            electrode_group=nwbfile.electrode_groups[
+                f"probe{probe_id}_shank{shank_num}"
+            ],
+            cluster_quality_label=unit["cluster_quality_label"],
+            spike_times=unit["spike_times"],
+            spike_depths=unit["spike_depths"],
+            waveform_mean=waveform_mean,
+        )
+
+    return units_table
+
+
+def add_ephys_units_to_nwb(
+    session_key: dict, nwbfile: pynwb.NWBFile, primary_clustering_paramset_idx: int = 0
+):
+    """
+    Add spiking data to NWBFile.
+
+    ephys.CuratedClustering.Unit::unit -> units.id
+    ephys.CuratedClustering.Unit::spike_times -> units["spike_times"]
+    ephys.CuratedClustering.Unit::spike_depths -> units["spike_depths"]
+    ephys.CuratedClustering.Unit::cluster_quality_label -> units["cluster_quality_label"]
+
+    ephys.WaveformSet.PeakWaveform::peak_electrode_waveform -> units["waveform_mean"]
+
+    Parameters
+    ----------
+    session_key: dict
+    nwbfile: pynwb.NWBFile
+    primary_clustering_paramset_idx: int, optional
+    """
+
+    if not ephys.ClusteringTask & session_key:
+        return
+
+    if nwbfile.electrodes is None:
+        add_electrodes_to_nwb(session_key, nwbfile)
+
+    # add additional columns to the units table
+    units_query = ephys.CuratedClustering.Unit() & session_key
+
+    for paramset_record in (
+        ephys.ClusteringParamSet & ephys.CuratedClustering & session_key
+    ).fetch("paramset_idx", "clustering_method", "paramset_desc", as_dict=True):
+        if paramset_record["paramset_idx"] == primary_clustering_paramset_idx:
+            units_table = create_units_table(
+                session_key,
+                nwbfile,
+                units_query,
+                paramset_record,
+                desc=paramset_record["paramset_desc"],
+            )
+            nwbfile.units = units_table
+        else:
+            name = f"units_{paramset_record['clustering_method']}"
+            units_table = create_units_table(
+                session_key,
+                nwbfile,
+                units_query,
+                paramset_record,
+                name=name,
+                desc=paramset_record["paramset_desc"],
+            )
+            ecephys_module = get_module(nwbfile, "ecephys")
+            ecephys_module.add(units_table)
+
+
+def get_electrodes_mapping(electrodes):
+    """
+    Create a mapping from the group (shank) and electrode id within that group to the row number of the electrodes
+    table. This is used in the construction of the DynamicTableRegion that indicates what rows of the electrodes
+    table correspond to the data in an ElectricalSeries.
+
+    Parameters
+    ----------
+    electrodes: hdmf.common.table.DynamicTable
+
+    Returns
+    -------
+    dict
+
+    """
+    return {
+        (electrodes["group"][idx].device.name, electrodes["id_in_probe"][idx],): idx
+        for idx in range(len(electrodes))
+    }
```

Comment:
I might be mistaken, but `electrodes["group"][idx].device.name` corresponds to the probe serial number (i.e. `probe_id`).  If what we want is the `shank_id`, we will need to parse `electrodes["group"][idx].name` which is constructed as follows: `name=f"probe{this_probe['probe']}_shank{shank_id}"`.

It does look like the `probe_id` is being used with the dictionary that is returned from this function.  In which case, this return statement is correctly and we would just need to update the doc string.  Thanks.

Reply:
Good catch. I think you are right that I want to use the shank ID number, not the probe ID number, and I didn't catch this because in this case each probe only has one shank.

Reply:
hmm now I am wondering if the docstring is just incorrect. I'll dig into this tomorrow. EIther way, something needs to be corrected.

Reply:
Or does this need to factor in the `probe_id` and the `shank_id`?

Reply:
the docstring was wrong

Reply:
Just to clarify, within the `add_electrodes_to_nwb` function the electrodes table is created for a given session for all probes and shanks.  So I would assume that the `get_electrodes_mapping` would need take into account both the probe and shank.

Reply:
This thread is resolved.  Since the `electrode` attribute in the `probe.ProbeType.Electrode` table is unique among all shanks in a probe, the mapping is correct as is.